base/namedtuple.jl

```diff
@@ -324,3 +326,38 @@ julia> Base.setindex(nt, "a", :a)
 function setindex(nt::NamedTuple, v, idx::Symbol)
     merge(nt, (; idx => v))
 end
+
+"""
+    @NamedTuple{key1::Type1, key2::Type2, ...}
+    @NamedTuple begin key1::Type1; key2::Type2; ...; end
+
+This macro gives a more convenient syntax for declaring `NamedTuple` types. It returns a `NamedTuple`
+type with the given keys and types, equivalent to `NamedTuple{(:key1, :key2, ...), Tuple{Type1,Type2,...}}`.
+If the `::Type` declaration is omitted, it is taken to be `Any`.   The `begin ... end` form allows the
+declarations to be split across multiple lines (similar to a `struct` declaration), but is otherwise
+equivalent.
+
+For example, the tuple `(a=3, b="hello")` has a type `NamedTuple{(:a, :b),Tuple{Int64,String}}`, which
+can also be declared via `@NamedTuple` as:
+
+```jldoctest
+julia> @NamedTuple{a::Int, b::String}
+NamedTuple{(:a, :b),Tuple{Int64,String}}
+
+julia> @NamedTuple begin
+           a::Int
+           b::String
+       end
+NamedTuple{(:a, :b),Tuple{Int64,String}}
+```
+"""
+macro NamedTuple(ex)
+    Meta.isexpr(ex, :braces) || Meta.isexpr(ex, :block) ||
+        throw(ArgumentError("@NamedTuple expects {...} or begin...end"))
+    decls = filter(e -> !(e isa LineNumberNode), ex.args)
+    all(e -> e isa Symbol || Meta.isexpr(e, :(::)), decls) ||
+        throw(ArgumentError("@NamedTuple must contain a sequence of name or name::type expressions"))
+    vars = [QuoteNode(e isa Symbol ? e : e.args[1]) for e in decls]
+    types = [e isa Symbol ? :Any : e.args[2] for e in decls]
```

Comment:
I guess you need `esc` to make it work for types defined outside `Base`?

```suggestion
    types = [esc(e isa Symbol ? :Any : e.args[2]) for e in decls]
```

Reply:
You don't need `esc` just to *read* external symbols, right?  Macro hygiene only rewrites variables that you assign to.

However, I guess we should put `esc` here out of paranoia that someone does something bananas like `NamedTuple{x::begin; someglobal += 1; Int}`. 😆